File: src/routes/timesheet.py
"""
routes/timesheet.py  —  Timesheet routes (Google Sheets backend)
══════════════════════════════════════════════════════════════════════════════

Routes in this file:
  GET  /api/timesheet               — Load all months of timesheet data
  POST /api/timesheet/update        — Save one cell after the user edits it
  POST /api/timesheet/add-month     — Create a new monthly tab in the Sheet
  POST /api/timesheet/delete-month  — Delete an empty monthly tab

How the timesheet is structured in Google Sheets:
  • One Google Sheet file contains all your timesheet data.
  • Each MONTH gets its own tab (called a "sheet") named "April 2026", "May 2026", etc.
  • Row 1 (the header row) has: "Activity" | "4/1/26" | "4/2/26" | … (one column per day)
  • Each subsequent row is one activity (e.g. "Tutoring", "Research") with hours per day.

  Visual example:
      Activity    | 4/1/26 | 4/2/26 | 4/3/26 | …
      Tutoring    |  2:00  |  3:00  |        | …
      Research    |  1:30  |        |  2:00  | …

A1 notation:
  Google Sheets cells are addressed like "B5" = column B, row 5.
  For columns beyond Z we use "AA", "AB", etc.  _col_letter() converts a
  1-based number into that letter string.
══════════════════════════════════════════════════════════════════════════════
"""
import logging
import calendar as cal_module
from datetime import datetime as dt

from flask import Blueprint, request, jsonify
from googleapiclient.discovery import build

from config import TIMESHEET_ID
from routes import require_auth

bp = Blueprint("timesheet", __name__)
logger = logging.getLogger(__name__)

# Month names indexed 0–11 (used to build tab titles like "April 2026")
MONTH_NAMES = [
    "January", "February", "March", "April", "May", "June",
    "July", "August", "September", "October", "November", "December",
]


def _col_letter(col: int) -> str:
    """Convert a 1-based column number into the A1-notation letter(s).

    Examples:
      1  → "A"
      26 → "Z"
      27 → "AA"
      28 → "AB"

    How it works:
      Each iteration peels off one letter from the right using divmod (division
      with remainder), then prepends it to the result string.
      This is similar to converting a decimal number to base-26.

    Args:
        col: 1-based column index (column A = 1, column B = 2, …)
    """
    result = ""
    while col > 0:
        col, rem = divmod(col - 1, 26)   # rem = 0–25 → maps to A–Z
        result = chr(65 + rem) + result  # chr(65) = 'A', chr(66) = 'B', …
    return result


@bp.route("/api/timesheet")
@require_auth
def api_timesheet(creds):
    """Load every month tab from the timesheet Google Sheet.

    Steps:
      1. Fetch the spreadsheet metadata to discover all tab names.
      2. Filter to only tabs named "Month YYYY" (e.g. "April 2026").
      3. For each month tab, read all rows and parse them into a structured format.
      4. Return the months sorted newest-first.

    Returns:
        {"months": [{"key": "2026-04", "label": "April 2026",
                     "dates": [...], "rows": [...]}, ...]}
    """
    try:
        sheets_svc = build("sheets", "v4", credentials=creds, cache_discovery=False)

        # Get spreadsheet metadata — this tells us what tabs exist
        meta = sheets_svc.spreadsheets().get(spreadsheetId=TIMESHEET_ID).execute()

        months = []
        for sheet in meta["sheets"]:
            # sheet["properties"]["title"] is the tab name, e.g. "April 2026"
            title = sheet["properties"]["title"].strip()
            parts = title.split()

            # Skip tabs that aren't in "Month YYYY" format
            if len(parts) != 2 or parts[0] not in MONTH_NAMES:
                continue
            try:
                year      = int(parts[1])
                month_num = MONTH_NAMES.index(parts[0]) + 1  # "April" → 4
            except ValueError:
                continue

            # "2026-04" — used to sort months chronologically
            month_key = f"{year}-{month_num:02d}"

            # Read all cell values from this tab
            result = sheets_svc.spreadsheets().values().get(
                spreadsheetId=TIMESHEET_ID,
                range=f"'{title}'",   # single quotes wrap the tab name in A1 notation
            ).execute()
            values = result.get("values", [])   # 2D list: values[row][col]

            if not values or len(values) < 2:
                continue   # tab has no data rows — skip

            headers = values[0]   # first row = ["Activity", "4/1/26", "4/2/26", …]

            # Parse each header column that looks like a date
            date_cols = []
            for col_i, h in enumerate(headers[1:], start=1):  # skip "Activity" col
                try:
                    raw = h.strip()
                    # Dates can be "4/1/26" (2-digit year) or "4/1/2026" (4-digit year)
                    date_obj = (
                        dt.strptime(raw, "%m/%d/%y")
                        if len(raw.split("/")[-1]) == 2
                        else dt.strptime(raw, "%m/%d/%Y")
                    )
                    date_cols.append({
                        "col_index": col_i,      # 0-based index in the row list
                        "sheet_col": col_i + 1,  # 1-based col number for A1 notation
                        "date_str":  raw,         # the raw string as it appears in the sheet
                        "date_obj":  date_obj,    # Python datetime for sorting
                    })
                except ValueError:
                    pass  # not a date header — skip it

            # Build one "row" object per activity (each row after the header)
            rows = []
            for row_i, row in enumerate(values[1:], start=1):
                activity = row[0] if row else ""
                if not activity:
                    continue  # blank rows in the sheet are skipped

                # Build a "cell" for each date column
                cells = []
                for dc in date_cols:
                    ci  = dc["col_index"]
                    val = row[ci] if ci < len(row) else ""  # empty if row is short
                    cells.append({
                        "value":     val,
                        "sheet_row": row_i + 1,    # +1 because header is row 1
                        "sheet_col": dc["sheet_col"],
                        "date":      dc["date_str"],
                    })
                rows.append({"activity": activity, "cells": cells})

            months.append({
                "key":        month_key,
                "label":      title,
                "sheet_name": title,
                "dates":      [dc["date_str"] for dc in date_cols],
                "rows":       rows,
            })

        # Sort months newest-first using the "YYYY-MM" key string
        months.sort(key=lambda m: m["key"], reverse=True)
        return jsonify({"months": months})

    except Exception as e:
        logger.exception("Timesheet load error")
        return jsonify({"error": str(e)}), 500


@bp.route("/api/timesheet/update", methods=["POST"])
@require_auth
def api_timesheet_update(creds):
    """Write a single edited cell value back to Google Sheets.

    This is called every time the user clicks away (blur event) after editing
    a time cell in the timesheet table.

    Expected JSON body:
        {"row": 3, "col": 5, "value": "2:30", "sheet_name": "April 2026"}

    A1 notation example:
        row=3, col=5 → "E3" (E is the 5th letter)
        With sheet name: "'April 2026'!E3"
    """
    try:
        body       = request.get_json()
        sheet_row  = int(body["row"])
        sheet_col  = int(body["col"])
        value      = body.get("value", "")
        sheet_name = body.get("sheet_name", "Sheet1")

        # Build the A1 cell reference, e.g. "'April 2026'!E3"
        cell_ref   = f"'{sheet_name}'!{_col_letter(sheet_col)}{sheet_row}"

        sheets_svc = build("sheets", "v4", credentials=creds, cache_discovery=False)
        sheets_svc.spreadsheets().values().update(
            spreadsheetId=TIMESHEET_ID,
            range=cell_ref,
            valueInputOption="USER_ENTERED",  # interprets "2:30" as a time value
            body={"values": [[value]]},       # 2D array: [[single cell value]]
        ).execute()
        logger.info("Saved %s → %s = %r", sheet_name, cell_ref, value)
        return jsonify({"success": True})

    except Exception as e:
        logger.warning("Timesheet update error: %s", e)
        return jsonify({"error": str(e)}), 500


@bp.route("/api/timesheet/add-month", methods=["POST"])
@require_auth
def api_timesheet_add_month(creds):
    """Create a new monthly tab in the Google Sheet.

    Steps:
      1. Figure out the tab title ("May 2026").
      2. Check the tab doesn't already exist.
      3. Copy the activity names (column A) from the most recent existing tab.
      4. Create the new tab (addSheet API call).
      5. Write the header row (Activity + one column per day) and activity rows.

    Expected JSON body:
        {"year": 2026, "month": 5}

    Returns:
        {"days_added": 31}  (the number of day columns created)
    """
    try:
        body      = request.get_json()
        year      = int(body["year"])
        month     = int(body["month"])
        tab_title = f"{MONTH_NAMES[month - 1]} {year}"   # e.g. "May 2026"

        sheets_svc      = build("sheets", "v4", credentials=creds, cache_discovery=False)
        meta            = sheets_svc.spreadsheets().get(spreadsheetId=TIMESHEET_ID).execute()
        existing_titles = [s["properties"]["title"] for s in meta["sheets"]]

        # Don't create a duplicate tab
        if tab_title in existing_titles:
            return jsonify({"days_added": 0, "message": "Tab already exists"})

        # ── Find activities from the most recent existing month tab ───────────
        month_tabs = []
        for s in meta["sheets"]:
            parts = s["properties"]["title"].strip().split()
            if len(parts) == 2 and parts[0] in MONTH_NAMES:
                try:
                    mn = MONTH_NAMES.index(parts[0]) + 1
                    yr = int(parts[1])
                    month_tabs.append((yr, mn, s["properties"]["title"]))
                except ValueError:
                    pass

        activity_list = []
        if month_tabs:
            month_tabs.sort(reverse=True)  # newest first
            source_tab = month_tabs[0][2]  # use the most recent tab as the template
            res        = sheets_svc.spreadsheets().values().get(
                spreadsheetId=TIMESHEET_ID,
                range=f"'{source_tab}'!A:A",  # column A only = activity names
            ).execute()
            col_a         = res.get("values", [])
            activity_list = [r[0].strip() for r in col_a[1:] if r and r[0].strip()]

        # ── Create the new empty tab ──────────────────────────────────────────
        sheets_svc.spreadsheets().batchUpdate(
            spreadsheetId=TIMESHEET_ID,
            body={"requests": [{"addSheet": {"properties": {"title": tab_title}}}]},
        ).execute()

        # ── Build and write the header + data rows ────────────────────────────
        # cal_module.monthrange(year, month)[1] returns the number of days in the month
        days_in_month = cal_module.monthrange(year, month)[1]

        # Date headers: "5/1/26", "5/2/26", … (2-digit year to match existing format)
        dates = [f"{month}/{day}/{str(year)[2:]}" for day in range(1, days_in_month + 1)]

        # First row: ["Activity", "5/1/26", "5/2/26", ...]
        # Each activity row:  ["Tutoring", "", "", ...]  (empty cells for all days)
        rows_to_write = [["Activity"] + dates]
        for act in activity_list:
            rows_to_write.append([act] + [""] * days_in_month)

        sheets_svc.spreadsheets().values().update(
            spreadsheetId=TIMESHEET_ID,
            range=f"'{tab_title}'!A1",
            valueInputOption="RAW",          # write strings as-is, no parsing
            body={"values": rows_to_write},
        ).execute()

        return jsonify({"days_added": days_in_month})

    except Exception as e:
        logger.exception("add-month error")
        return jsonify({"error": str(e)}), 500


@bp.route("/api/timesheet/delete-month", methods=["POST"])
@require_auth
def api_timesheet_delete_month(creds):
    """Delete an entire monthly tab from the Google Sheet.

    The UI only shows the Delete button on months where ALL cells are empty,
    so this is safe — you can't accidentally delete filled-in data via the UI.

    Expected JSON body:
        {"month_key": "April 2026"}   ← the tab's display name

    Steps:
      1. Find the tab's internal numeric sheetId (different from its title).
      2. Call deleteSheet with that ID.
    """
    try:
        body      = request.get_json()
        tab_title = body["month_key"]   # the tab display name, e.g. "April 2026"

        sheets_svc = build("sheets", "v4", credentials=creds, cache_discovery=False)
        meta       = sheets_svc.spreadsheets().get(spreadsheetId=TIMESHEET_ID).execute()

        # Find the internal sheetId for the tab with this title
        sheet_id = next(
            (s["properties"]["sheetId"] for s in meta["sheets"]
             if s["properties"]["title"] == tab_title),
            None,
        )
        if sheet_id is None:
            return jsonify({"error": "Tab not found"}), 404

        # Delete the tab — this is permanent!
        sheets_svc.spreadsheets().batchUpdate(
            spreadsheetId=TIMESHEET_ID,
            body={"requests": [{"deleteSheet": {"sheetId": sheet_id}}]},
        ).execute()
        return jsonify({"success": True})

    except Exception as e:
        logger.exception("delete-month error")
        return jsonify({"error": str(e)}), 500

File: static/js/timesheet.js
// ── timesheet.js ──────────────────────────────────────────────────────────────
//
// Manages the Timesheet tab — a view into a Google Sheets spreadsheet where
// you record hours worked on different activities each day.
//
// How it's structured:
//   • The spreadsheet has one tab per month: "April 2026", "May 2026", etc.
//   • Each tab has a header row of dates and rows of activities with hour values.
//   • The dashboard loads all months from the server and renders them as
//     collapsible <details> sections.
//   • Long months (>7 days shown at once) are split into weekly sub-tabs.
//   • Each cell is editable directly — clicking it, typing a value, and
//     clicking away saves the change back to Google Sheets.

let _timesheetLoaded = false;
let _nextMonth       = null;   // {year, month} — computed after load, used by "Add Month" button


// ── formatTimeDisplay ─────────────────────────────────────────────────────────
// Normalises a raw cell value into a consistent time format for display.
// Examples:
//   "2"   → "2:00"    (bare number = hours)
//   "30m" → "0:30"    (30 minutes)
//   "1:30"→ "1:30"    (already correct, left unchanged)
//   ""    → ""        (empty cell, left empty)
function formatTimeDisplay(val) {
  if (!val) return '';
  // Already in "H:MM" format — return as-is
  if (/^\d+:\d{2}$/.test(String(val).trim())) return String(val).trim();
  // Strip trailing "h" or "m" suffix, parse as a number
  const n = parseFloat(String(val).replace(/[hm]$/i, '').trim());
  if (isNaN(n)) return String(val);   // not a recognisable number — return raw
  if (n >= 0  && n <= 9)  return n + ':00';           // whole hours 0–9
  if (n >= 10 && n <= 59) return '0:' + String(n).padStart(2, '0');  // minutes
  return String(val);
}


// ── formatDateShort ───────────────────────────────────────────────────────────
// Converts a date string like "4/18/25" into a short label like "Apr 18"
// for use in the table column headers.
function formatDateShort(dateStr) {
  try {
    const parts  = dateStr.split('/');
    const months = ['Jan','Feb','Mar','Apr','May','Jun','Jul','Aug','Sep','Oct','Nov','Dec'];
    return `${months[parseInt(parts[0])-1]} ${parseInt(parts[1])}`;
  } catch { return dateStr; }
}


// ── loadTimesheet ─────────────────────────────────────────────────────────────
// Fetches all month data from the server (/api/timesheet) and renders it.
// Shows a loading message while waiting, or an error if the fetch fails.
async function loadTimesheet() {
  const container = document.getElementById('timesheet-container');
  container.innerHTML = '<div class="loading">Loading timesheet...</div>';
  try {
    const res  = await fetch(`${SERVER}/api/timesheet`);
    const data = await res.json();
    if (data.error) { container.innerHTML = `<div class="empty">Error: ${data.error}</div>`; return; }
    renderTimesheet(data);
    updateAddMonthBtn(data.months);
    _timesheetLoaded = true;
  } catch(e) {
    container.innerHTML = '<div class="empty">Could not load timesheet — make sure Google is connected.</div>';
  }
}


// ── renderTimesheet ───────────────────────────────────────────────────────────
// Builds and injects the full timesheet HTML from the server data.
// Each month becomes a <details> element (click the header to expand/collapse).
// Wide months are split into weekly sub-tabs so the table doesn't scroll horizontally.
function renderTimesheet(data) {
  const container = document.getElementById('timesheet-container');
  if (!data.months || !data.months.length) {
    container.innerHTML = '<div class="empty">No timesheet data found.</div>';
    return;
  }

  // ── makeWeekTable ──────────────────────────────────────────────────────────
  // Returns the HTML for one week's worth of data as a <table>.
  // weekDates  — the date strings for this week (column headers)
  // startIdx   — which cell index to start reading from in each row's cells[]
  function makeWeekTable(month, weekDates, startIdx) {
    return `
      <table class="timesheet-table">
        <thead><tr>
          <th class="ts-act-header">Activity</th>
          ${weekDates.map(d => `<th>${formatDateShort(d)}</th>`).join('')}
        </tr></thead>
        <tbody>
          ${month.rows.map(row => `
            <tr>
              <td class="ts-act-cell">${row.activity}</td>
              ${row.cells.slice(startIdx, startIdx + weekDates.length).map(cell => `
                <td><span
                  class="ts-cell"
                  contenteditable="true"
                  data-row="${cell.sheet_row}"
                  data-col="${cell.sheet_col}"
                  data-sheet="${month.sheet_name}"
                  data-original="${formatTimeDisplay(cell.value)}"
                  onblur="onTsCellBlur(this)"
                  onkeydown="if(event.key==='Enter'){event.preventDefault();this.blur();return false}"
                >${formatTimeDisplay(cell.value)}</span></td>
              `).join('')}
            </tr>
          `).join('')}
        </tbody>
      </table>`;
  }

  let html = '';
  data.months.forEach((month, mi) => {
    // First month is open by default; the rest start collapsed
    const isOpen  = mi === 0 ? 'open' : '';
    // A "Delete" button appears only on completely empty months (safety guard)
    const isEmpty = month.rows.every(r => r.cells.every(c => !c.value));
    const deleteBtn = isEmpty
      ? `<button onclick="event.preventDefault();deleteTimesheetMonth('${month.sheet_name}','${month.label}')"
           style="margin-left:auto;background:rgba(247,106,106,0.12);color:var(--accent4);border:1px solid rgba(247,106,106,0.3);border-radius:0.3rem;padding:0.15rem 0.6rem;font-size:0.73rem;cursor:pointer">Delete</button>`
      : '';

    // Split dates into chunks of 7 (one chunk per week tab)
    const weeks = [];
    for (let i = 0; i < month.dates.length; i += 7) {
      weeks.push({ dates: month.dates.slice(i, i + 7), startIdx: i });
    }

    // mKey = safe CSS id using underscores instead of hyphens
    const mKey   = month.key.replace(/-/g, '_');
    // Tab bar only appears if the month has more than one week of data
    const tabBar = weeks.length > 1 ? `
      <div class="ts-week-tabs">
        ${weeks.map((w, wi) => {
          const first = parseInt(w.dates[0].split('/')[1]);
          const last  = parseInt(w.dates[w.dates.length - 1].split('/')[1]);
          return `<button class="ts-week-tab ${wi === 0 ? 'active' : ''}" onclick="switchTsWeek('${mKey}',${wi})">${first}–${last}</button>`;
        }).join('')}
      </div>` : '';

    const panes = weeks.map((w, wi) => `
      <div class="ts-week-pane ${wi === 0 ? 'active' : ''}">
        ${makeWeekTable(month, w.dates, w.startIdx)}
      </div>`).join('');

    html += `<details class="timesheet-month card" ${isOpen}>
      <summary><span class="ts-arrow">▶</span>${month.label} <span style="color:var(--muted);font-size:0.78rem;font-weight:400;margin-left:0.4rem">${month.dates.length} days</span>${deleteBtn}</summary>
      <div class="timesheet-table-wrap" id="ts-month-${mKey}">
        ${tabBar}
        ${panes}
      </div>
    </details>`;
  });

  container.innerHTML = html;

  // ── Smooth close animation ─────────────────────────────────────────────────
  // By default <details> closes instantly.  We intercept the click, play a
  // slide-up animation, THEN remove the "open" attribute.
  container.querySelectorAll('details.timesheet-month').forEach(details => {
    details.querySelector('summary').addEventListener('click', e => {
      if (!details.open) return;   // already closed — let it open normally
      e.preventDefault();          // stop the default instant-close
      const content = details.querySelector('.timesheet-table-wrap');
      if (!content) { details.removeAttribute('open'); return; }
      content.style.animation = 'tsClose 0.32s cubic-bezier(0.4,0,0.2,1) forwards';
      content.addEventListener('animationend', () => {
        details.removeAttribute('open');
        content.style.animation = '';
      }, { once: true });   // { once: true } removes the listener after it fires
    });
  });
}


// ── switchTsWeek ──────────────────────────────────────────────────────────────
// Shows a specific week pane and highlights its tab button.
// Called by the week tab buttons rendered inside each month section.
function switchTsWeek(mKey, weekIdx) {
  const wrap = document.getElementById('ts-month-' + mKey);
  if (!wrap) return;
  // Toggle the "active" class on tab buttons and panes
  wrap.querySelectorAll('.ts-week-tab').forEach((t, i)  => t.classList.toggle('active', i === weekIdx));
  wrap.querySelectorAll('.ts-week-pane').forEach((p, i) => p.classList.toggle('active', i === weekIdx));
}


// ── _tsSaveCell ───────────────────────────────────────────────────────────────
// Makes one attempt to POST a cell value to the server.
// Returns true on success, false on any failure (HTTP error or network error).
async function _tsSaveCell(sheetRow, sheetCol, sheetName, value) {
  try {
    const res = await fetch(`${SERVER}/api/timesheet/update`, {
      method:  'POST',
      headers: { 'Content-Type': 'application/json' },
      body:    JSON.stringify({ row: sheetRow, col: sheetCol, value, sheet_name: sheetName }),
    });
    if (!res.ok) {
      console.warn('Timesheet save HTTP', res.status, { sheetRow, sheetCol, sheetName });
      return false;
    }
    const json = await res.json();
    if (!json.success) {
      console.warn('Timesheet save server error:', json.error, { sheetRow, sheetCol, sheetName });
      return false;
    }
    return true;
  } catch(e) {
    console.warn('Timesheet save network error:', e);
    return false;
  }
}


// ── onTsCellBlur ──────────────────────────────────────────────────────────────
// Called when the user clicks away from an editable timesheet cell (onblur).
// Formats the value, compares it to the original, and saves if it changed.
// Retries once after 4 seconds if the first attempt fails (handles transient 502s).
async function onTsCellBlur(el) {
  const raw      = el.textContent.trim();
  const stripped = /^\d+:\d{2}$/.test(raw) ? raw : raw.replace(/[hm]$/i, '').trim();
  const formatted = formatTimeDisplay(stripped);

  if (el.textContent.trim() !== formatted) el.textContent = formatted;

  const original = el.dataset.original || '';
  if (formatted === original) return;

  const sheetRow  = parseInt(el.dataset.row);
  const sheetCol  = parseInt(el.dataset.col);
  const sheetName = el.dataset.sheet;

  el.classList.add('saving');
  el.classList.remove('save-error');

  let ok = await _tsSaveCell(sheetRow, sheetCol, sheetName, formatted);

  if (!ok) {
    // First attempt failed — wait 4 s then retry once
    await new Promise(r => setTimeout(r, 4000));
    ok = await _tsSaveCell(sheetRow, sheetCol, sheetName, formatted);
  }

  if (ok) {
    el.dataset.original = formatted;
  } else {
    el.classList.add('save-error');
    console.error('Timesheet save failed after retry', { sheetRow, sheetCol, sheetName, formatted });
  }
  el.classList.remove('saving');
}


// ── updateAddMonthBtn ─────────────────────────────────────────────────────────
// Sets the "Add Next Month" button label after loading the timesheet.
// Looks at the most recent existing month and computes the one after it.
// e.g. if the newest tab is "April 2026", the button says "Add May 2026".
function updateAddMonthBtn(months) {
  if (!months || !months.length) return;
  const [y, m] = months[0].key.split('-').map(Number);
  let ny = y, nm = m + 1;
  if (nm > 12) { nm = 1; ny++; }
  const labels = ['January','February','March','April','May','June','July','August','September','October','November','December'];
  _nextMonth = { year: ny, month: nm };
  const btn  = document.getElementById('ts-add-btn');
  if (btn) btn.textContent = `Add ${labels[nm-1]} ${ny}`;
}


// ── addTimesheetMonth ─────────────────────────────────────────────────────────
// Sends a request to the server to create the next month's tab in Google Sheets.
// The server copies activity names from the most recent tab and builds the header row.
async function addTimesheetMonth() {
  const status = document.getElementById('ts-add-status');
  if (!_nextMonth) { status.textContent = 'Load timesheet first.'; return; }
  status.textContent = 'Adding…';
  try {
    const res  = await fetch(`${SERVER}/api/timesheet/add-month`, {
      method:  'POST',
      headers: { 'Content-Type': 'application/json' },
      body:    JSON.stringify(_nextMonth),
    });
    const json = await res.json();
    if (json.error) { status.textContent = 'Error: ' + json.error; return; }
    status.textContent = `✓ Added ${json.days_added} days`;
    setTimeout(() => { status.textContent = ''; }, 3000);
    loadTimesheet();   // reload to show the new month
  } catch(e) {
    status.textContent = 'Failed to add month.';
  }
}


// ── deleteTimesheetMonth ──────────────────────────────────────────────────────
// Asks the user for confirmation then permanently deletes the month's tab.
// The server only allows this on empty months (checked in renderTimesheet).
async function deleteTimesheetMonth(monthKey, label) {
  if (!confirm(`Delete all ${label} columns from the sheet? This cannot be undone.`)) return;
  try {
    const res  = await fetch(`${SERVER}/api/timesheet/delete-month`, {
      method:  'POST',
      headers: { 'Content-Type': 'application/json' },
      body:    JSON.stringify({ month_key: monthKey }),
    });
    const json = await res.json();
    if (json.error) { alert('Error: ' + json.error); return; }
    loadTimesheet();
  } catch(e) { alert('Failed to delete month.'); }
}
